src/main/java/org/folio/edge/sip2/repositories/PatronRepository.java

```diff
@@ -0,0 +1,354 @@
+        "Users repository cannot be null");
+    this.circulationRepository = Objects.requireNonNull(circulationRepository,
+        "Circulation repository cannot be null");
+    this.clock = Objects.requireNonNull(clock, "Clock cannot be null");
+  }
+
+  /**
+   * Perform patron information.
+   *
+   * @param patronInformation the patron information domain object
+   * @return the patron information response domain object
+   */
+  public Future<PatronInformationResponse> patronInformation(
+      PatronInformation patronInformation,
+      SessionData sessionData) {
+    Objects.requireNonNull(patronInformation, "patronInformation cannot be null");
+    Objects.requireNonNull(sessionData, "sessionData cannot be null");
+
+    final String barcode = patronInformation.getPatronIdentifier();
+
+    // Look up the patron by barcode
+    final Future<JsonObject> result = usersRepository.getUserByBarcode(barcode, sessionData);
+    return result.compose(user -> {
+      if (user == null || !user.getBoolean("active", FALSE).booleanValue()) {
+        return invalidPatron(patronInformation);
+      } else {
+        final String userId = user.getString("id");
+        if (userId == null) {
+          // Something is really messed up if the id is missing
+          log.error("User with barcode {} is missing the \"id\" field", barcode);
+          return invalidPatron(patronInformation);
+        }
+        return validPatron(userId, user, patronInformation, sessionData);
+      }
+    });
+  }
+
+  private Future<PatronInformationResponse> validPatron(String userId, JsonObject user,
+      PatronInformation patronInformation, SessionData sessionData) {
+    // Now that we have a valid patron, we can retrieve data from circulation
+    final PatronInformationResponseBuilder builder = PatronInformationResponse.builder();
+    addPersonalData(user, builder);
+    final Integer startItem = patronInformation.getStartItem();
+    final Integer endItem = patronInformation.getEndItem();
+    final Future<PatronInformationResponseBuilder> holdsFuture = circulationRepository
+        .getRequestsByUserId(userId, "Hold", startItem, endItem, sessionData).map(
+            holds -> addHolds(holds, patronInformation.getSummary() == HOLD_ITEMS, builder));
+    final Future<PatronInformationResponseBuilder> overdueFuture =
+        circulationRepository.getOverdueLoansByUserId(userId, ZonedDateTime.now(clock),
+            startItem, endItem, sessionData).map(
+                overdues -> addOverdueItems(overdues,
+                    patronInformation.getSummary() == OVERDUE_ITEMS, builder));
+    final Future<PatronInformationResponseBuilder> recallsFuture =
+        getRecalls(userId, sessionData).compose(recalls -> addRecalls(recalls, startItem, endItem,
+            patronInformation.getSummary() == RECALL_ITEMS, builder));
+    return CompositeFuture.all(holdsFuture, overdueFuture, recallsFuture)
```

Comment:
More lambda magic... How do the results of holds/overdue/recallsFutures getting used by the builder?

Reply:
In this case, there is not too much magic. I'll add some comments to describe the workflow here, but basically, each async call has the same builder and adds the items it is responsible for to the builder, holds (count and items, if requested), overdues (same as holds) and recalls (same as holds, but assembled differently). When all the futures complete (`CompositeFuture.all()`), we add the final items and build the response.

Because of the complexity (over 60) of the `patronInformation` method I had to split it up quite a bit, which actually simplified some of the logic that was in there. It is still a bit complicated.